pygmt/clib/session.py

```diff
@@ -1098,8 +1158,12 @@ def virtualfile_from_vectors(self, *vectors):
             family, geometry, mode="GMT_CONTAINER_ONLY", dim=[columns, rows, 1, 0]
         )
 
-        for col, array in enumerate(arrays):
+        # Use put_vector for first n columns with numerical type data
+        for col, array in enumerate(arrays[:columns]):
             self.put_vector(dataset, column=col, vector=array)
+        # Use put_strings for last column with string type data
+        for array in arrays[columns:]:
+            self.put_strings(dataset, family="GMT_IS_VECTOR", strings=array)
```

Comment:
These codes are incorrect. We can't call `GMT_Put_Strings` multiple times.

See the comments in https://github.com/GenericMappingTools/pygmt/issues/483#issuecomment-647251367, we have to combine all trailing string arrays into one array and call `GMT_Put_Strings` once.

Reply:
It does just call put strings once, or zero times. If you examine the code before, the slice either returns a list with one item or a list with zero items.

Reply:
Do you mean these lines?

```python
        columns = len(arrays)
        if np.issubdtype(arrays[-1].dtype, np.str_):
            columns -= 1
```

It won't work for input with multiple trailing string arrays. For example, for the `text` module, the input array may be
```
x = [1, 2]
y = [3, 4]
font = ["15p,1,blue", "15p,1,red"]
text = ["labelA", "labelB"]
```
We have to combine `font` and `text` arrays into one single string array before passing it via GMT_Put_Strings.
```
newarray = ["15p,1,blue labelA", "15p,1,red labelB"]
```

Reply:
Ok I see, but do we want to the concatenation:

1. **Before** passing it into `lib.virtualfile_from_vectors`, i.e. at the `fig.text` (and also `fig.meca`, etc) level
2. **In** the clib itself, i.e. find all the arrays that have a 'str' data type, and concatenate them inside `lib.virtualfile_from_vectors` before passing it into `put_strings`.

Reply:
Option 2 is easier and more intuitive for module wrapper developers.

Reply:
Not so easy for the `clib` developer though :laughing:. I do agree with going with Option 2, but I'm not sure how the memory pointers would work if we concatenate two numpy arrays with 'str' dtypes. It might be that we'll be duplicating memory somehow.

Reply:
From what I now about Python and GMT, I think we may have to duplicate the memory in PyGMT, and then duplicate it again in GMT.